packages/contracts/src/registry/SpaceRegistry.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.8.17;
+
+import { OwnableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { DAOFactory } from "@aragon/osx/framework/dao/DAOFactory.sol";
+import { DAO } from "@aragon/osx/core/dao/DAO.sol";
+
+import { ISpaceRegistry } from "./ISpaceRegistry.sol";
+
+contract SpaceRegistry is OwnableUpgradeable, UUPSUpgradeable,ISpaceRegistry {
+
+    DAOFactory public daoFactory;
+    mapping(bytes16 => address) public daoAddressBySpaceId;
+    mapping(address => bytes16) public spacesByDAOAddress;
+    mapping(address => bytes16) public mainPersonalSpaceByAddress;
+    mapping(address => address) public pendingMainPersonalSpaceDAOByAddress;
+
+    event SpaceRegistryInitialized(address daoFactory, address owner);
+    event SpaceRegistrySpaceCreated(bytes16 indexed spaceId, address indexed dao, address indexed creator);
+    event SpaceRegistryMainPersonalSpaceUpdatePending(address indexed user, bytes16 indexed spaceId, address indexed dao);
+    event SpaceRegistryMainPersonalSpaceSet(address indexed user, bytes16 indexed previousSpaceId, bytes16 indexed newSpaceId);
+
+    error SpaceRegistryInvalidZeroAddress();
+    error SpaceRegistryInvalidSpaceId(bytes16 spaceId);
+    error SpaceRegistryInvalidCaller(address caller);
+
+    function initialize(address _owner, address _daoFactory) external initializer {
+        if(_daoFactory == address(0)) revert SpaceRegistryInvalidZeroAddress();
+
+        __Ownable_init();
+        _transferOwnership(_owner);
+        daoFactory = DAOFactory(_daoFactory);
+
+        emit SpaceRegistryInitialized(_daoFactory, _owner);
+    }
+
+    function createSpace(DAOFactory.DAOSettings calldata _daoSettings, DAOFactory.PluginSettings[] calldata _pluginSettings, bool _isMainPersonalSpace) external override returns (DAO createdDao, bytes16 spaceId) {
+        createdDao = daoFactory.createDao(_daoSettings, _pluginSettings);
+        spaceId = generateSpaceId(address(createdDao));
+        daoAddressBySpaceId[spaceId] = address(createdDao);
+        spacesByDAOAddress[address(createdDao)] = spaceId;
+
+        if(_isMainPersonalSpace) {
+            bytes16 previousMainPersonalSpace = mainPersonalSpaceByAddress[msg.sender];
+            mainPersonalSpaceByAddress[msg.sender] = spaceId;
+            emit SpaceRegistryMainPersonalSpaceSet(msg.sender, previousMainPersonalSpace, spaceId);
+        }
+
+        emit SpaceRegistrySpaceCreated(spaceId, address(createdDao), msg.sender);
+    }
+
+    function setMainPersonalSpace(bytes16 _spaceId) external override {
+        address daoAddress = daoAddressBySpaceId[_spaceId];
+        if (daoAddress == address(0)) revert SpaceRegistryInvalidSpaceId(_spaceId);
+
+        pendingMainPersonalSpaceDAOByAddress[msg.sender] = daoAddress;
```

Comment:
With this implementation the mechanics of setting a personal space would only work with spaces created by the factory, correct? Not sure if it makes sense to point to _any_ valid deployment (on any chain), even if it's not created by the factory.

Reply:
Yes, the idea is that all spaces should be created through this registry, otherwise you could point to invalid addresses or contracts that are not a valid DAO... I was considering the multi-chain aspect as out of scope for this MVP. For a multi-chain registry we can either:
- Deploy a space registry contract like this one on each chain, and define rules to resolve conflicts at the indexing layer (e.g. if a user sets their main personal space on several chains, only treat the last one as valid), or
- Use an interoperability layer to connect space registry contracts on each chain, so that they communicate with the one on the Geo chain to keep a single source of truth.
But I would recommend tackling this in a second iteration.